benchmarks/benches/room_list_bench.rs

```diff
@@ -0,0 +1,176 @@
+use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion, Throughput};
+use futures_util::{pin_mut, StreamExt};
+use matrix_sdk::test_utils::logged_in_client_with_server;
+use matrix_sdk_ui::{room_list_service::filters::new_filter_non_left, RoomListService};
+use serde::Deserialize;
+use serde_json::{json, Map, Value};
+use tokio::runtime::Builder;
+use wiremock::{http::Method, Match, Mock, Request, ResponseTemplate};
+
+struct SlidingSyncMatcher;
+
+impl Match for SlidingSyncMatcher {
+    fn matches(&self, request: &Request) -> bool {
+        request.url.path() == "/_matrix/client/unstable/org.matrix.msc3575/sync"
+            && request.method == Method::POST
+    }
+}
+
+#[derive(Deserialize)]
+pub(crate) struct PartialSlidingSyncRequest {
+    pub txn_id: Option<String>,
+}
+
+fn criterion() -> Criterion {
+    #[cfg(target_os = "linux")]
+    let criterion = Criterion::default().with_profiler(pprof::criterion::PProfProfiler::new(
+        100,
+        pprof::criterion::Output::Flamegraph(None),
+    ));
+
+    #[cfg(not(target_os = "linux"))]
+    let criterion = Criterion::default();
+
+    criterion
+}
+
+pub fn room_list(c: &mut Criterion) {
+    let runtime = Builder::new_multi_thread()
+        .enable_time()
+        .enable_io()
+        .build()
+        .expect("Can't create runtime");
+
+    let mut benchmark = c.benchmark_group("room_list");
+
+    for number_of_rooms in [10, 100, 1000, 10_000] {
+        benchmark.throughput(Throughput::Elements(number_of_rooms));
+        benchmark.bench_with_input(
+            BenchmarkId::new("sync", number_of_rooms),
+            &number_of_rooms,
+            |benchmark, maximum_number_of_rooms| {
+                benchmark.to_async(&runtime).iter(|| async {
```

Comment:
There are a lot of line in this benchmark that should be considered to be part of the setup. I don't think we want to time `wiremock` starting up.

Since there are so many lines that are timed that shouldn't be, I think it's quite hard to get the benchmark to settle down. I.e. after some 10 runs I finally got it to settle down, but still it's quite the noisy benchmark:

```
    time:   [-2.6332% +7.1045% +17.983%] (p = 0.18 > 0.05)
    thrpt:  [-15.242% -6.6332% +2.7044%]
    No change in performance detected.
```

Please take a look at the [`Bencher`](https://docs.rs/criterion/latest/criterion/struct.Bencher.html) documentation and more specifically at the [`iter_batched()`](https://docs.rs/criterion/latest/criterion/struct.AsyncBencher.html#method.iter_batched) method:

> If your routine requires some per-iteration setup that shouldn’t be timed, use iter_batched or iter_batched_ref.

Having that many lines as part of the benchmark also makes it quite hard to see what exactly we're measuring. So what are we intending to measure? The sync response processing, the room list creation?